clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -7205,27 +7205,65 @@ static void handleSYCLKernelAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
     const NamedDecl *TParam = TL->getParam(I);
     if (isa<NonTypeTemplateParmDecl>(TParam)) {
       S.Diag(FT->getLocation(),
-             diag::warn_sycl_kernel_invalid_template_param_type);
+             diag::warn_sycl_attr_invalid_template_param_type)
+          << "sycl_kernel";
       return;
     }
   }
 
   // Function must have at least one argument.
   if (getFunctionOrMethodNumParams(D) != 1) {
-    S.Diag(FT->getLocation(), diag::warn_sycl_kernel_num_of_function_params);
+    S.Diag(FT->getLocation(), diag::warn_sycl_attr_num_of_function_params)
+        << "sycl_kernel";
     return;
   }
 
   // Function must return void.
   QualType RetTy = getFunctionOrMethodResultType(D);
   if (!RetTy->isVoidType()) {
-    S.Diag(FT->getLocation(), diag::warn_sycl_kernel_return_type);
+    S.Diag(FT->getLocation(), diag::warn_sycl_attr_return_type)
+        << "sycl_kernel";
     return;
   }
 
   handleSimpleAttribute<SYCLKernelAttr>(S, D, AL);
 }
 
+static void handleSYCLKernelImplAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+  // The 'sycl_kernel_impl' attribute applies only to function templates.
+  const auto *FD = cast<FunctionDecl>(D);
+  const FunctionTemplateDecl *FT = FD->getDescribedFunctionTemplate();
+  assert(FT && "Function template is expected");
+
+  // Template parameters must be typenames.
+  const TemplateParameterList *TL = FT->getTemplateParameters();
+  for (unsigned I = 0; I < TL->size(); ++I) {
+    const NamedDecl *TParam = TL->getParam(I);
+    if (isa<NonTypeTemplateParmDecl>(TParam)) {
+      S.Diag(FT->getLocation(),
+             diag::warn_sycl_attr_invalid_template_param_type)
+          << "sycl_kernel_impl";
+      return;
+    }
+  }
+
+  // Function must have at least one argument.
+  if (getFunctionOrMethodNumParams(D) < 1) {
+    S.Diag(FT->getLocation(), diag::warn_sycl_attr_num_of_function_params)
+        << "sycl_kernel_impl";
+    return;
+  }
+
+  // Function must return void.
+  if (!getFunctionOrMethodResultType(D)->isVoidType()) {
+    S.Diag(FT->getLocation(), diag::warn_sycl_attr_return_type)
+        << "sycl_kernel_impl";
```

Comment:
Why do you need additional attribute?
`sycl_kernel_impl` looks fully identical to `sycl_kernel` attribute. Can we just use `sycl_kernel` attribute instead?

Reply:
The main issue is the restriction on the number of template arguments and the number of function arguments on a routine marked "sycl_kernel".   The implementation routines that I have seen seem to need a bit more leeway than that.  Can that restriction on the "sycl_kernel" routines be lifted?  If so, we can reuse the same attribute.

Reply:
I meant "parameters" in both places that I said "arguments" above.

Reply:
Those restrictions are not coming from the spec, they aim to guarantee that the right function is marked with this attribute, so it can be used to construct OpenCL kernel.
We can lift some of them if we are sure that OpenCL kernel generation won't be broken.

Reply:
Thanks for this information, @Fznamznon!  Since OpenCL kernel generation happens on the device, I will try to reuse the same attribute on the host for this PR purpose.